src/utils/astFromValue.js

```diff
@@ -0,0 +1,129 @@
+/* @flow */
+/**
+ *  Copyright (c) 2015, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import invariant from '../utils/invariant';
+import isNullish from '../utils/isNullish';
+import type { Value } from '../language/ast';
+import {
+  INT,
+  FLOAT,
+  STRING,
+  BOOLEAN,
+  ENUM,
+  LIST,
+  OBJECT,
+  OBJECT_FIELD,
+} from '../language/kinds';
+import {
+  GraphQLType,
+  GraphQLEnumType,
+  GraphQLInputObjectType,
+  GraphQLList,
+  GraphQLNonNull,
+} from '../type/definition';
+import { GraphQLFloat } from '../type/scalars';
+
+
+/**
+ * Produces a GraphQL Value AST given a JavaScript value.
+ *
+ * Optionally, a GraphQL type may be provided, which will be used to
+ * disambiguate between value primitives.
+ *
+ * | JSON Value    | GraphQL Value        |
+ * | ------------- | -------------------- |
+ * | Object        | Input Object         |
+ * | Array         | List                 |
+ * | Boolean       | Boolean              |
+ * | String        | String / Enum Value  |
+ * | Number        | Int / Float          |
+ *
+ */
+export default function astFromValue(
+  value: any,
+  type?: ?GraphQLType
+): ?Value {
+  if (type instanceof GraphQLNonNull) {
+    // Note: we're not checking that the result is non-null.
+    // This function is not responsible for validating the input value.
+    return astFromValue(value, type.ofType);
+  }
+
+  if (isNullish(value)) {
+    return null;
+  }
+
+  // Convert JavaScript array to GraphQL list. If the GraphQLType is a list, but
+  // the value is not an array, convert the value using the list's item type.
+  if (Array.isArray(value)) {
+    var itemType = type instanceof GraphQLList ? type.ofType : null;
+    return {
+      kind: LIST,
+      values: value.map(item => astFromValue(item, itemType))
+    };
+  } else if (type instanceof GraphQLList) {
```

Comment:
This is quite confusing to me. Is this because arguments have the list coercion rules?